apps/src/javalab/redux/editorRedux.ts

```diff
@@ -0,0 +1,320 @@
+/**
+ * Redux store for editor-specific Java Lab state.
+ */
+
+// TODO: Can we fix our imports and no longer need to ignore this rule?
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+const {
+  fileMetadataForEditor,
+  updateAllSourceFileOrders
+} = require('@cdo/apps/javalab/JavalabFileHelper');
+import _ from 'lodash';
+import {createSelector, createSlice, PayloadAction} from '@reduxjs/toolkit';
+import {JavalabEditorDialog} from '../types';
+
+interface Sources {
+  [key: string]: {
+    text: string;
+    tabOrder: number;
+    isVisible?: boolean;
+    isValidation?: boolean;
+  };
+}
+
+interface FileMetadata {
+  [key: string]: string;
+}
+
+// JavalabEditorDialog is an enum of possible open dialogs.
+// keyof typeof gives us the union type of the all the Enum keys as strings.
+// https://www.typescriptlang.org/docs/handbook/enums.html#enums-at-compile-time
+type JavlabEditorDialogOptions = keyof typeof JavalabEditorDialog;
+
+interface JavalabEditorState {
+  fileMetadata: FileMetadata;
+  orderedTabKeys: string[];
+  activeTabKey: string;
+  lastTabKeyIndex: number;
+  sources: Sources;
+  validation: Sources;
+  editorOpenDialogName: JavlabEditorDialogOptions | null;
```

Comment:
Curious why you use `null` here but `undefined` in [prior PR](https://github.com/code-dot-org/code-dot-org/pull/51246#discussion_r1162682886) for interface JavalabState.

Reply:
Generally I went with whatever we were using before--in this one, the default value for `editorOpenDialogName` was `null`, but in the previous one, the default value was `undefined`. The main difference is `levelName` should be set to something pretty early on, and then left alone, but `editorOpenDialogName` will be expected to switch between a value in the enum and `null`. I don't know if we had a great heuristic here to start with though.